### ISSUE TITLE
Initial flicker when using ProvideWindowInsets

### ISSUE THREAD

**Issue by chrisbanes** [google/accompanist#155]
I think the issue is that we only request insets once a composition has happened. The starting values are all `0` the initial layout of your composables will be as if there were no insets.

On the first composition we request insets, then need to wait for the next view layout pass (for dispatch), then the next composition to update the right values, then the new Compose layout.

AFAICT This is mostly an issue for apps which use `Fragment`s and then `ProvideWindowInsets` in each, since all the above happens on each fragment view creation. We should investigate if there is a way for us to provide an API which allows fragments to 'prime' the values earlier.

Raised by @apkelly

**Comment by apkelly** [comment]
Just to confirm, the app I'm working on does use Fragments, each one provides its UI using Compose, we still use the existing Jetpack Navigation via Fragments for navigating between Fragments in a single Activity app.

**Comment by chrisbanes** [comment]
Worked this out. So using `ProvideWindowInsets` in fragments will result in the following:

1. Fragment is added.
2. `onCreateView` is called.
3. Fragment's `ComposeView` is added to the hierarchy.
4. Some time later (_after_ initial view layout pass AFAICT), the first composition happens. 
    - `ProvideWindowInsets` is called which sets listeners and requests an insets pass.
5. Compose content is laid out using `0` sized insets.
6. In the next view layout pass, insets are dispatched and the Compose `WindowInsets` are updated.
7. Compose content is laid out using correct insets.

The visual flicker is caused by steps 5 & 7 happens quickly after each other. 

Using `observeFromView()` from #158 instead results in the following:

1. Fragment is added
2. `onCreateView` is called
    - `observeFromView()` is called from `onCreateView()` which sets listeners and requests an insets pass.
3. Fragment's `ComposeView` is added to the hierarchy.
4. In the first layout of `ComposeView` insets are dispatched, and this library's `WindowInsets` are updated.
5. Compose content is composed using correct insets from first layout.

No flicker! 😅

**Comment by apkelly** [comment]
#shipit!

**Comment by Nghicv** [comment]
@apkelly @chrisbanes I still got flicker when applying `observeFromView`. I am using fragments with navigation.

**Comment by chrisbanes** [comment]
@Nghicv do you have an example of this? Does it happen on specific devices. I'm not sure there's much else we can do.

**Comment by Nghicv** [comment]
@chrisbanes I have just tested TV app from the google play store, it got flicker too
this is my device
<img width="438" alt="Screen Shot 2021-02-08 at 09 23 19" src="https://user-images.githubusercontent.com/9463238/107169128-b9309080-69ef-11eb-9f13-01297d18f2fc.png">

**Comment by radzio** [comment]
@apkelly  @chrisbanes  

have you managed to make it work with `FragmentContainerView` and the Navigation Library (multiple fragments with composable UI) ? 

I have one Activity 

```
class HomeActivity : FragmentActivity() {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)

        binding = ActivityMainBinding.inflate(layoutInflater)

        setContentView(binding.root)
    }
  ```

and fragments like: 

```
class MyFragment: BaseFragment() {

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = ComposeView(requireContext()).apply {
        layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, MATCH_PARENT)

        val windowInsets = ViewWindowInsetObserver(this).start(consumeWindowInsets = false)

        setContent {
            Providers(
                LocalWindowInsets provides windowInsets
            ) {
                // composable UI 
            }
        }
    }
```

my UI is flickering on Pixel 2 Emulator API 30 and OnePlus 8T API 30

**Comment by chrisbanes** [comment]
@radzio if you have a simple project, would you mind uploading it? (zip is fine). I've got a few apps which use fragments and I haven't seen any flickers.

**Comment by radzio** [comment]
@chrisbanes 

Maybe we are talking about different bug/feature but check this video: 

https://user-images.githubusercontent.com/469111/108363715-89ef0000-71f5-11eb-9758-86230fec3975.mp4

"Hi there text" evidently starts without inset padding (just play it in slow-mo). Maybe I'm doing something wrong but I have no idea what :(.

When I'm not using insets and `statusBarsPadding` modifier there is no jumping at all. TextField is, as expected, in wrong place ,under status bar): 
<img src="https://user-images.githubusercontent.com/469111/108364407-53fe4b80-71f6-11eb-8b0b-5f3cdd688c49.png" width="20%" />

Example app: 
[Flickering.zip](https://github.com/chrisbanes/accompanist/files/6003071/Flickering.zip)

**Comment by chrisbanes** [comment]
Thanks for the project. I finally managed to reproduce this (note to self: change the emulator to use 1 CPU core). Now to work out what's happening....

**Comment by chrisbanes** [comment]
OK, got a little further with this. This is what happens internally:

```
D/ComposeView Parent: Attached to window
D/ComposeView: Attached to window

D/ComposeView Parent: onMeasure

# First measure/layout of Composables. WindowInsets are empty
D/InsetsPaddingModifier: measure(): Insets(left=0, top=0, right=0, bottom=0)

D/ComposeView Parent: onMeasure
D/ViewWindowInsetObserver: WindowInsets dispatched and received

D/ComposeView Parent: onMeasure
D/ComposeView Parent: onMeasure

D/ComposeView Parent: onLayout
D/ComposeView: First layout

# Second measure/layout of Composables. WindowInsets are now filled
D/InsetsPaddingModifier: measure(): Insets(left=0, top=48, right=0, bottom=0)
```

As a summary: the composables are laid out once before the `WindowInsets` are dispatched. I'm not sure there's much we can do here.

**Comment by radzio** [comment]
@chrisbanes  thanks for the investigation. It's sad that we can't do more here. 

However, at least in my case, I think that I can use `activity.window.decorView.rootWindowInsets` 

On the first fragment, it will not work because it's null, but it's ok for me because the first fragment is not inset-dependent.  But on the subsequent fragments, I get this value immediately. However, I don't know how to use this `rootWindowInsets` because I can't set anything (setters are internal  :) ) 

BTW do you think that the Jetpack Compose team can do something about this? Maybe something needs to be changed in ComposeView?

**Comment by chrisbanes** [comment]
> BTW do you think that the Jetpack Compose team can do something about this? Maybe something needs to be changed in ComposeView?

Yeah I've reached out and hoping  we can come up with something.

I'm still investigating why the `WindowInsets` aren't dispatched in the first measure pass. If we could fix that, we wouldn't have any isseues.

**Comment by chrisbanes** [comment]
I just come up with a bit of a hack which fixes this. It's safe to use, but I don't want to ship it in the library until I speak to some of the team: https://gist.github.com/chrisbanes/bd9aa241dbab971c23433aaff8104d38.

**Comment by radzio** [comment]
Thanks! Works like a charm :)

**Comment by Nghicv** [comment]
@chrisbanes @radzio the hack worked but I got a white screen before the destination screen is shown. seem to Time to showing white screen equal to the time to wait MAXIMUM_ONMEASURE_IGNORED

**Comment by chrisbanes** [comment]
@Nghicv that would make sense, as that is what the hack is doing.

**Comment by Nghicv** [comment]
@chrisbanes should we come up with a better solution?

**Comment by natario1** [comment]
Is this related to fragments only? We're seeing the same jump in an app without fragments, not sure if I should open a separate issue or investigate more.

**Comment by svenjacobs** [comment]
> I just come up with a bit of a hack which fixes this. It's safe to use, but I don't want to ship it in the library until I speak to some of the team: https://gist.github.com/chrisbanes/bd9aa241dbab971c23433aaff8104d38, but it should unblock you all.

Thanks! This hack works for me (with a slight modification: `Providers` -> `CompositionLocalProvider`), however is there now a more reliable and stable solution?

**Comment by alexvanyo** [comment]
Some experimental insets APIs have been merged into `androidx`, available in the latest snapshots (if you're feeling adventurous, feel free to try it out and report issues or feedback!)

The mechanism behind the implementation is a little bit different than `accompanist`, and in my initial testing the approach seems to be more reliable around flickering and in fragments.

**Comment by mxkmn** [comment]
So, seems like I found how to fix this: https://stackoverflow.com/q/77807932/12544067
